src/librustc_trans/trans/adt.rs

```diff
@@ -917,10 +946,9 @@ pub fn trans_get_discr<'blk, 'tcx>(bcx: Block<'blk, 'tcx>, r: &Repr<'tcx>,
                        range_assert)
         }
         Univariant(..) => C_u8(bcx.ccx(), 0),
-        RawNullablePointer { nndiscr, nnty, .. } =>  {
-            let cmp = if nndiscr == Disr(0) { IntEQ } else { IntNE };
-            let llptrty = type_of::sizing_type_of(bcx.ccx(), nnty);
-            ICmp(bcx, cmp, Load(bcx, scrutinee), C_null(llptrty), DebugLoc::None)
+        RawForbiddenValue { payload_discr, forbidden_value, .. } =>  {
+            let cmp = if payload_discr == Disr(0) { IntEQ } else { IntNE };
+            ICmp(bcx, cmp, Load(bcx, scrutinee), forbidden_value, DebugLoc::None)
```

Comment:
(is it just me, or had this code always been over the threshold of "wow there's a lot comparisons and double-negation-like trickery just to map from the question of "are you the forbidden value?" to the goal of extracting desired discriminant value itself?)

In other words: If you happen to feel like trying to come up with a one or two line comment explaining the reasoning here, I won't mind. :)